sycl/include/sycl/ext/intel/esimd/math.hpp

```diff
@@ -1176,128 +1175,140 @@ bfn(T src0, T src1, T src2) {
 
 /// @} sycl_esimd_logical
 
-/// Performs add with carry of 2 unsigned 32-bit vectors.
+/// Performs add with carry of 2 unsigned integral vectors.
 /// @tparam N size of the vectors
 /// @param carry vector that is going to hold resulting carry flag
 /// @param src0 first term
 /// @param src1 second term
 /// @return sum of 2 terms, carry flag is returned through \c carry parameter
-template <int N>
-__ESIMD_API __ESIMD_NS::simd<uint32_t, N>
-addc(__ESIMD_NS::simd<uint32_t, N> &carry, __ESIMD_NS::simd<uint32_t, N> src0,
-     __ESIMD_NS::simd<uint32_t, N> src1) {
-  std::pair<__ESIMD_DNS::vector_type_t<uint32_t, N>,
-            __ESIMD_DNS::vector_type_t<uint32_t, N>>
-      Result = __esimd_addc<uint32_t, N>(src0.data(), src1.data());
-
-  carry = Result.first;
-  return Result.second;
+template <int N, typename T>
+__ESIMD_API std::enable_if_t<!std::is_signed_v<T> && std::is_integral_v<T>,
+                             __ESIMD_NS::simd<T, N>>
+addc(__ESIMD_NS::simd<T, N> &carry, __ESIMD_NS::simd<T, N> src0,
+     __ESIMD_NS::simd<T, N> src1) {
+#ifdef __SYCL_DEVICE_ONLY__
+  std::pair<__ESIMD_DNS::vector_type_t<T, N>, __ESIMD_DNS::vector_type_t<T, N>>
+      Result = __spirv_IAddCarry<T, N>(src0.data(), src1.data());
+
+  carry = Result.second;
+  return Result.first;
+#else
+  return 0;
+#endif // __SYCL_DEVICE_ONLY__
 }
 
-/// Performs add with carry of a unsigned 32-bit vector and scalar.
+/// Performs add with carry of a unsigned integral vector and scalar.
 /// @tparam N size of the vectors
 /// @param carry vector that is going to hold resulting carry flag
 /// @param src0 first term
 /// @param src1 second term
 /// @return sum of 2 terms, carry flag is returned through \c carry parameter
-template <int N>
-__ESIMD_API __ESIMD_NS::simd<uint32_t, N>
-addc(__ESIMD_NS::simd<uint32_t, N> &carry, __ESIMD_NS::simd<uint32_t, N> src0,
-     uint32_t src1) {
-  __ESIMD_NS::simd<uint32_t, N> Src1V = src1;
+template <int N, typename T>
```

Comment:
I believe the help of a macro would make it better: a) there will be no need for SFINAE and b) the templates would not require extra parameter `T`, which goes in very unusual order - after `N` - (usual order is `<T, N>`). Finally, passing `simd_view` as the 1 arg to vector+scalar op will require specifying only `N`, instead of `<N, T>`.
The idea is: (it may need some adaptation because of using __SYCL_DEVICE_ONLY__ inside it:
```C++
#define __ESIMD_ADDC(T) \
template <int N> \
__ESIMD_API __ESIMD_NS::simd<T, N>> \
addc(__ESIMD_NS::simd<T, N> &carry, __ESIMD_NS::simd<T, N> src0, \
     __ESIMD_NS::simd<T, N> src1) { \
#ifdef __SYCL_DEVICE_ONLY__ \
  std::pair<__ESIMD_DNS::vector_type_t<T, N>, __ESIMD_DNS::vector_type_t<T, N>> \
      Result = __spirv_IAddCarry<T, N>(src0.data(), src1.data()); \
  carry = Result.second; \
  return Result.first; \
#else \
  return 0; \
#endif // __SYCL_DEVICE_ONLY__ \
} \
/* OTHER 3 variants here (vec+scalar, scalar+vec, sccalar+scalar  */ \

__ESIMD_ADDC(uint8_t)
__ESIMD_ADDC(uint16_t)
__ESIMD_ADDC(uint32_t)
__ESIMD_ADDC(uint64_t)

```